boards: arm: stm32h7b3i_dk: Enable CAN peripheral and transceiver

tmilkovic51 opened the pull request:
- Add CAN transceiver to board DTS
- Enable CAN peripheral and set correct pinctrl config in board DTS
- Specify in board yaml file that board supports CAN and CANFD
- Modify board documentation

Fixes: #45790

tmilkovic51 committed:
boards: arm: stm32h7b3i_dk: Enable CAN peripheral and transceiver

Add CAN transceiver to board DTS
Enable CAN peripheral and set correct pinctrl config in board DTS
Add CAN support in board yaml file
Modify board documentation

Signed-off-by: Tomislav Milkovic <tomislav.milkovic95@gmail.com>

tmilkovic51 commented on boards/arm/stm32h7b3i_dk/doc/index.rst:
```diff
@@ -87,14 +89,18 @@ Default Zephyr Peripheral Mapping:
     - SDCLK/SDNWE/SDCKE1/SDNE1 : PG8/PH5/PH7/PH6
+
 - LTDC :
+
```
These newlines fix background highlighthing problem
![image](https://user-images.githubusercontent.com/33724303/168623752-6fc3c7e9-cdb9-4708-a2e6-77d1450a6b90.png)

henrikbrixandersen commented on boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts:
```diff
@@ -49,6 +50,13 @@
 		zephyr,memory-region = "SDRAM2";
 	};
 
+	can_phy: can_transceiver {
```
```suggestion
	transceiver0: can-phy0 {
```

henrikbrixandersen commented on boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts:
```diff
@@ -62,17 +70,18 @@
 	status = "okay";
 };
 
+/* PLL1P is used for system clock (280 MHz), PLL1Q is used for FDCAN bit quantum clock (16 MHz) */
```
16MHz seems a bit low. Does it pass all the CAN timing test? Have you tried with 80MHz (div-q = 7?) instead?

tmilkovic51 replied:
Thanks for the quick review, this clock definitely needs changing. 
I have only tested on CAN2.0 compatible baudrates since I don't have a CANFD compatible adapter (I used PCAN-USB adapter and BUSMASTER on Windows).

Would testing with CAN controller in loopback mode (like in `samples/drivers/can` sample) be good enough?

tmilkovic51 replied:
One little note about timing tests - I ran them with modified min timings in `can_stm32h7.c` because these assertions in `can_mcan.c` fail:
`
void can_mcan_configure_timing(struct can_mcan_reg *can,
			       const struct can_timing *timing,
			       const struct can_timing *timing_data)
{
	if (timing) {
		uint32_t nbtp_sjw = can->nbtp & CAN_MCAN_NBTP_NSJW_MSK;

		__ASSERT_NO_MSG(timing->prop_seg == 0);
		__ASSERT_NO_MSG(timing->phase_seg1 <= 0x100 &&
				timing->phase_seg1 > 0);
		__ASSERT_NO_MSG(timing->phase_seg2 <= 0x80 &&
				timing->phase_seg2 > 0);
		__ASSERT_NO_MSG(timing->prescaler <= 0x200 &&
				timing->prescaler > 0);
		__ASSERT_NO_MSG(timing->sjw == CAN_SJW_NO_CHANGE ||
				(timing->sjw <= 0x80 && timing->sjw > 0));

		can->nbtp = (((uint32_t)timing->phase_seg1 - 1UL) & 0xFF) <<
				CAN_MCAN_NBTP_NTSEG1_POS |
			    (((uint32_t)timing->phase_seg2 - 1UL) & 0x7F) <<
				CAN_MCAN_NBTP_NTSEG2_POS |
			    (((uint32_t)timing->prescaler  - 1UL) & 0x1FF) <<
				CAN_MCAN_NBTP_NBRP_POS;
`
This is because min timings config for `phase_seg1`, `phase_seg2` and `prescaler` in `can_stm32h7.c` are configured to `0x00`:
`
.timing_min = {
	.sjw = 0x00,
	.prop_seg = 0x00,
	.phase_seg1 = 0x00,
	.phase_seg2 = 0x00,
	.prescaler = 0x00
}
`
They should be configured to 0x01, since `can_mcan` decrements them by 1 before writing them to NBTP register.

Do I fix it in this PR or should I open another one?

tmilkovic51 replied:
@henrikbrixandersen Thanks for the thorough review.
I added the fix for issue described above in a separate commit, please take a look if it's OK.

IMO, having `phase_seg1` and `phase_seg2` set to 0 doesn't make sense (where is the sampling point in that case?), it should be an invalid configuration (and it is rightfully checked in `can_mcan.c`). The driver first decrements the value by 1 and writes it to register, but hardware interprets the value as: value_in_register + 1.

henrikbrixandersen replied:
Indeed. Good catch. Please see my question about sjw min value.

henrikbrixandersen commented on boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts:
```diff
@@ -128,6 +137,20 @@
+	pinctrl-0 = <&fdcan1_rx_pa11 &fdcan1_tx_pa12>;
+	pinctrl-names = "default";
+	phys = <&can_phy>;
+	bus-speed = <500000>;
+	bus-speed-data = <500000>;
+	sjw = <1>;
+	sjw-data = <1>;
+	sample-point = <875>;
+	sample-point-data = <875>;
```
```suggestion
	bus-speed = <125000>;
	bus-speed-data = <1000000>;
	sample-point = <875>;
	sample-point-data = <875>;
```

tmilkovic51 commented:
In `samples/drivers/can/prj.conf` file, there is an assignment for Kconfig option `CONFIG_CAN_MAX_FILTER=5`. Since STM32H7 CANFD driver doesn't have `CAN_MAX_FILTER` option defined in its Kconfig tree (it has separate options for max number of standard and extended filters), Kconfig throws a warning. 
Is there any way to "delete" a Kconfig define in Kconfig overlay file which would be specific for this board?

henrikbrixandersen commented:
> In `samples/drivers/can/prj.conf` file, there is an assignment for Kconfig option `CONFIG_CAN_MAX_FILTER=5`. Since STM32H7 CANFD driver doesn't have `CAN_MAX_FILTER` option defined in its Kconfig tree (it has separate options for max number of standard and extended filters), Kconfig throws a warning. 
> 
> Is there any way to "delete" a Kconfig define in Kconfig overlay file which would be specific for this board?

No. Please ignore the warning for now.

tmilkovic51 committed:
drivers: can: can_stm32h7: Fix min timings

Assertions in can_mcan failed because the driver expects values
strictly greater than 0

Signed-off-by: Tomislav Milkovic <tomislav.milkovic95@gmail.com>

henrikbrixandersen commented on boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts:
```diff
@@ -49,6 +50,13 @@
 	};
 
+	transceiver0: can-phy0 {
+		compatible = "can-transceiver-gpio";
```
```suggestion
		compatible = "microchip,mcp2562fd", "can-transceiver-gpio";
```

tmilkovic51 replied:
Added

henrikbrixandersen reviewed (changes requested):
One final nit: please include the CAN transceiver make/model in the compatible list.

henrikbrixandersen commented on drivers/can/can_stm32h7.c:
```diff
@@ -115,12 +115,12 @@ static const struct can_driver_api can_stm32h7_driver_api = {
 	.timing_min = {
 		.sjw = 0x00,
```
This likely goes for sjw as well?

```suggestion
		.sjw = 0x01,
```

tmilkovic51 replied:
You're right, it does.
With value 0x00 (`CAN_SJW_NO_CHANGE`) the MCAN driver retains the previous SJW value that was written in `NBTP`/`DBTP` register, which after reset is 0x03, which is not the minimal configurable value.

Changed.

henrikbrixandersen commented:
@erwango It would be nice to get this into v3.1.0 if possible - otherwise I think we need to split out the timing fix to a separate PR?

erwango commented on boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts:
```diff
@@ -49,6 +50,13 @@
 		zephyr,memory-region = "SDRAM2";
 	};
 
+	transceiver0: can-phy0 {
+		compatible = "microchip,mcp2562fd", "can-transceiver-gpio";
+		standby-gpios = <&gpioh 8 GPIO_ACTIVE_HIGH>;
+		max-bitrate = <5000000>;
+		#phy-cells = <0>;
+	};
```
I'm not finding this as part of BSP. Shouldn't it be moved to a sample overlay ?

erwango replied:
ok, found as "MC**D**2562FD" in the user manual.
Well, as long as it is clearly documented, which is the case, I'm fine.

erwango reviewed (commented):
One point to be confirmed, otherwise looks good
